src/common/cockpitconf.c

```diff
@@ -21,282 +21,387 @@
+  ssize_t line_len;
+
+  cockpit_conf_loaded = true;
+
+  f = fopen (file_path, "r");
+  if (!f)
+    {
+      if (errno != ENOENT)
+        warnx ("couldn't load configuration file: %s: %m\n", file_path);
+      return false;
+    }
+
+  regcompx (&re_section, "^[[:space:]]*\\[([^][[:cntrl:]]+)\\][[:space:]]*$", REG_EXTENDED|REG_NEWLINE);
+  regcompx (&re_keyval, "^[[:space:]]*([[:alnum:]-]+)[[:space:]]*=[[:space:]]*(.*)$", REG_EXTENDED|REG_NEWLINE);
+  regcompx (&re_ignore, "^[[:space:]]*(#.*)?$", REG_EXTENDED|REG_NOSUB);
+
+  errno = 0;
+  while ((line_len = getline (&line, &line_size, f)) >= 0)
     {
-      gchar **keys;
-      section = g_hash_table_lookup (cockpit_conf, groups[i]);
-      if (section == NULL)
+      /* maximum number of () matches that we want to capture from the above REs, + 1 for the entire string (group 0) */
+      const int MAX_MATCH = 3;
+      regmatch_t matches[MAX_MATCH];
+
+      if (regexec (&re_section, line, MAX_MATCH, matches, 0) == 0)
         {
-          section = g_hash_table_new_full (cockpit_str_case_hash, cockpit_str_case_equal, g_free, g_free);
-          g_hash_table_insert (cockpit_conf, g_strdup (groups[i]), section);
+          free (cur_section);
+          cur_section = strndupx (line + matches[1].rm_so, matches[1].rm_eo - matches[1].rm_so);
         }
 
-      keys = g_key_file_get_keys (key_file, groups[i], NULL, &error);
-      g_return_val_if_fail (error == NULL, FALSE);
+      else if (regexec (&re_keyval, line, 3, matches, 0) == 0)
+        {
+          Entry *e;
+
+          if (!cur_section)
+            {
+              warnx ("%s: key=val line not in any section: %s", file_path, line);
+              ret = false;
+              break;
+            }
+
+          e = malloc (sizeof (Entry));
+          if (e == NULL)
+            err (1, "failed to malloc");
+          e->section = strdupx (cur_section);
+          e->key = strndupx (line + matches[1].rm_so, matches[1].rm_eo - matches[1].rm_so);
+          e->value = strndupx (line + matches[2].rm_so, matches[2].rm_eo - matches[2].rm_so);
+          e->strv_value = NULL;
+          e->strv_cache = NULL;
+          /* prepend new Entry to cockpit_conf; that way, later values win over earlier ones in a forward search */
+          e->next = cockpit_conf;
+          cockpit_conf = e;
+        }
 
-      for (j = 0; keys[j] != NULL; j++)
+      else if (regexec (&re_ignore, line, 0, NULL, 0) == 0)
         {
-          gchar *value = g_key_file_get_value (key_file, groups[i], keys[j], &error);
-          g_return_val_if_fail (error == NULL, FALSE);
-          g_hash_table_insert (section, g_strdup (keys[j]), value);
+          /* comment or empty line */
+        }
+      else
+        {
+          warnx ("%s: invalid line: %s", file_path, line);
+          ret = false;
+          break;
         }
-      g_strfreev (keys);
     }
-  g_strfreev (groups);
-
-  g_debug ("Loaded configuration from: %s", file_path);
+  if (errno != 0)
```

Comment:
The `errno = 0;` at the top of the loop isn't good enough.  We could have `errno` set from (for example) an out-of-bounds `strtoll()` call on the last iteration of the loop, and we don't reset it before calling `getline()` again.

Reply:
Fixed, thank you!